crypto/ex_data.c

```diff
@@ -287,7 +287,7 @@ int CRYPTO_dup_ex_data(int class_index, CRYPTO_EX_DATA *to,
         CRYPTOerr(CRYPTO_F_CRYPTO_DUP_EX_DATA, ERR_R_MALLOC_FAILURE);
         return 0;
     }
-    if (!CRYPTO_set_ex_data(to, mx - 1, NULL))
+    if (!CRYPTO_set_ex_data(to, mx - 1, CRYPTO_get_ex_data(to, mx - 1)))
```

Comment:
You'd have to be very careful about how you implement your` dup_func`, since it's not just simply `strdup`. But that's neither here nor there.

What this is doing is making sure that there are at least `mx` elements in the ex_data stack, by setting the expected last element to itself (source is `to`, destination is `to`). This is really a no-op, but has the side-effect of ensuring that the number of elements in the ex_data stack is equal to `mx`. If the `mx - 1` element does not exist (unlikely, but possible), `CRYPTO_get_ex_data()` returns `NULL`, so that element is set to `NULL`, which is fine.

That being said, if it there was allocated ex_data at index `mx - 1`, this would set it to NULL and subsequently leak memory.